% =============================================================================
% File:  sample_slides.tex --  Example of the use of the Falkor Beamer theme
% Author(s): Sebastien Varrette <Sebastien.Varrette@uni.lu>
% Time-stamp: <Mar 2014-04-29 16:06 svarrette>
% 
% Copyright (c) 2012 Sebastien Varrette <Sebastien.Varrette@uni.lu>
% .             http://varrette.gforge.uni.lu
% 
% For more information:
% - LaTeX: http://www.latex-project.org/
% - Beamer: https://bitbucket.org/rivanvx/beamer/
% - LaTeX symbol list:
% http://www.ctan.org/tex-archive/info/symbols/comprehensive/symbols-a4.pdf
% 
% Latest version of these files can be found on Github:
% 

% =============================================================================
\documentclass{beamer}
% \documentclass[draft]{beamer}
\usepackage{_style}
\usepackage{__config}

% The key part to use my theme
\usetheme{Falkor}

% Not integrated in my theme as not everybody wants that
\AtBeginSection[]
{
  \frame{
    \frametitle{Summary}
    {\scriptsize\tableofcontents[currentsection]}
  }
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%% Header %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\title{\EventName}
\subtitle{\TPindex: \TPtitle}

\author{\authors}
\institute[UL]{
  University of Luxembourg, Luxembourg
}

% Mandatory to define a logo - otherwise compilation will fail in an unobvious
% manner.
\pgfdeclareimage[height=0.8cm]{logo}{images/logo_UL.pdf}
%\logo{\pgfuseimage{logo}}
\date{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% Body %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{document}

\begin{frame}
    \vspace{2.5em}
    \titlepage
\end{frame}

% .......
\frame{
  \begin{center}
      \textbf{Latest versions available on
        \href{https://github.com/ULHPC/}{Github}}:
      \vfill
      \begin{description}
        \item[UL HPC tutorials:] \hfill
          \myurl{https://github.com/ULHPC/tutorials}
        \item[UL HPC School:] \hfill
          \myurl{http://hpc.uni.lu/hpc-school/}
        \item[\TPindex tutorial sources:] \hfill
          \myurl{\TPghurl}
      \end{description}
  \end{center}
}

% ......
\frame{
  \frametitle{Summary}
  {\scriptsize
    \tableofcontents
  }
}

% ===============================================
\section{Objectives}

% ............
\frame{
  \frametitle{Objective of this PS}

  Better understand the usage of Bioinformatics packages on the \ULHPC.
  \hfill \\ \hfill \\

  \pause   Why Bioinformatics? 3Vs:
  \begin{itemize}
    \pause \item very relevant in the context of the UL/LCSB
    \pause \item very fast growing domain
    \pause \item very many associated workflows, thus excellent examples
  \end{itemize}

  \pause \includegraphics[width=0.2\linewidth,keepaspectratio]{graphics/abyss-logo.png}
  \pause \includegraphics[width=0.2\linewidth,keepaspectratio]{graphics/bowtie-logo.png}
  \pause \includegraphics[width=0.2\linewidth,keepaspectratio]{graphics/tophat-logo.png}
  \pause \includegraphics[width=0.2\linewidth,keepaspectratio]{graphics/GROMACS-logo.png}
  \pause \includegraphics[width=0.2\linewidth,keepaspectratio]{graphics/mpiblast-logo.png}
  
}

\section{Bioinformatics packages}

% .......
\frame{
  \frametitle{ABySS}
  
  \textbf{ABySS}: Assembly By Short Sequences \\
  \hfill \scriptsize{a de novo, parallel, paired-end sequence assembler designed for short reads}
  \\ \hfill \\ \hfill \\
  \pause
  \begin{itemize}
    \item several applications in the ABySS package
    \item only \textbf{ABYSS-P} is parallelized using MPI \\
    $ \hookrightarrow $ started with the \textbf{abyss-pe} launcher
    \item workflow (pipeline) of \textbf{abyss-pe} also includes:
    \begin{itemize}
     \item OpenMP-parallel applications
     \item serial applications
    \end{itemize}
    \item Note: { \scriptsize compared with other de novo assemblers, the per-node memory requirements are smaller due to ABySS' task distribution model}
  \end{itemize} 
}

% .......
\frame{
  \frametitle{Gromacs}
  
  \textbf{GROMACS}: GROningen MAchine for Chemical Simulations \\
  \hfill \scriptsize{versatile package for molecular dynamics, primarily designed for biochemical molecules}
  \\ \hfill \\
  \pause
  \begin{itemize}
    \item very large codebase: 1.836.917 SLOC
    \item many applications in the package, several parallelization modes
    \item \textbf{mdrun}: computational chemistry engine, performing:\\
      {
      \scriptsize
      $ \hookrightarrow $ molecular dynamics simulations \\
      $ \hookrightarrow $ Brownian Dynamics, Langevin Dynamics \\
      $ \hookrightarrow $ Conjugate Gradient \\
      $ \hookrightarrow $ L-BFGS \\
      $ \hookrightarrow $ Steepest Descents energy minimization \\
      $ \hookrightarrow $ Normal Mode Analysis \\
      }
    \item \textbf{mdrun} - parallelized using MPI, OpenMP, pthreads and with support for GPU acceleration
  \end{itemize}
}
 
% .......
\frame{
  \frametitle{Bowtie2/TopHat}

  \textbf{Bowtie2}: Fast and sensitive read alignment \\
  \hfill { \scriptsize ultrafast \& memory-efficient alignment of sequencing reads to long ref. sequences} \\

  \textbf{TopHat}: A fast spliced read mapper for RNA-Seq \\
  \hfill { \scriptsize alignment of RNA-Seq reads to a genome, to identify exon-exon splice junctions}
  \hfill \\  \hfill \\
  \pause
  \begin{itemize}
    \item TopHat aligns reads to mammalian-sized genomes using Bowtie
    \item then analyzes the mapping results to identify splice junctions between exons
    \item \textbf{bowtie2} is OpenMP-parallel
    \item rest of workflow is sequential
  \end{itemize} 
}

% .......
\frame{
  \frametitle{mpiBLAST}
  
  \textbf{mpiBLAST}: Open-Source Parallel BLAST \\
  \hfill { \scriptsize parallel implementation of NCBI BLAST, scaling to hundreds of processors }
\pause  
\begin{itemize}
 \item two main applications: \textbf{mpiblast} \textbf{mpiformatdb}
 \item requires (NCBI) substitution matrices and formatted BLAST databases
 \item the databases can be segmented
 \begin{itemize}
  \item into as many segments as the \# of cores that will be used when performing searches
  \item or a multiple, in order to avoid load imbalance
 \end{itemize}

 \item \textbf{mpiblast} requires >= 3 processes, 2 used for internal tasks \\
 $ \hookrightarrow $ mpirun -np 3 mpiblast $[...]$ only gives you one searcher process! \\
\end{itemize} 
}

\section{Notes}
\frame{
  \frametitle{Notes..}
  .. on real world applications (bioinfo or others):
    \begin{itemize}
      \item make sure you \textit{understand the parallel capabilities} of your software \\
      {\scriptsize $ \hookrightarrow $ pthreads/OpenMP vs MPI vs hybrid} \\
      {\scriptsize $ \hookrightarrow $ use of GPU acceleration} \\
     \pause \item make sure you \textit{request the appropriate resources} for the processing needs of your workflow\\
      \begin{itemize}
       \item {\scriptsize Does the software always take advantage of more than 1 core or node?}
       \item {\scriptsize How does it scale? Many obstacles to perfect scalability!}
      \end{itemize}

    \end{itemize}
   \hfill \\ 
   
  \pause .. on data management:
    \begin{itemize}
      \item make sure you use \textit{the appropriate storage place} \\
      {\scriptsize $ \hookrightarrow $ \$HOME vs \$WORK vs \$SCRATCH }
      \item stage data in/out, archive your (many \& unused) 'small' files
    \end{itemize}  
}


\section{Practical session}

% .......
\frame{
  \frametitle{Exercises}
    \begin{itemize}
      \item Read and understand the Bioinformatics tutorial \myurl{\TPghurl} \\
      \item Run the examples \\
      $ \hookrightarrow $ all calculations should be fast \\
      $ \hookrightarrow $ you should attempt the exercises proposed in each section \\
      \item Try even more tests, e.g.:
      \begin{itemize}
       \item on different node classes
       \item with one core per node on >= 2 nodes
       \item ~ vs >= 2 cores on single node
      \end{itemize}
    \end{itemize}
    

}

\section{Conclusion}

% ............
\begin{frame}
  \frametitle{Conclusion}
  
  \begin{itemize}
    \item Bioinformatics applications execution on the \ULHPC
    \item Outlined:
    \begin{itemize}
      \item different workflows
      \item some of the concepts you should care about when running complex software
    \end{itemize}
  \end{itemize}

    \begin{block}{Perspectives}
        \begin{itemize}
          \item Personalize the UL HPC launchers with the specific commands for ABySS, Gromacs, TopHat, Bowtie, mpiBLAST..
        \end{itemize}
    \end{block}

\end{frame}

\section*{Thank you for your attention...}
\frame{
  \frametitle{Questions?}
  \begin{center}
      \includegraphics[scale=0.2]{question.jpg}
  \end{center}

  {\tiny
    \tableofcontents

  }
}

\newcounter{finalframe}
\setcounter{finalframe}{\value{framenumber}}

% \appendix

% \frame{
%   \frametitle{Appendix}
% 
%   \begin{acronym}\setlength\itemsep{-0.3em}
%       \acro{DFT}{Discrete Fourier Transform}
%       \acro{EA}{Evolutionary Algorithm}
%       \acro{PRNG}{[Pseudo]-Random Number Generator}
%       \acro{UL}{University of Luxembourg}
%   \end{acronym}
% 
% }

\setcounter{framenumber}{\value{finalframe}}

\end{document}

% ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
% eof
% 
% Local Variables:
% mode: latex
% mode: flyspell
% mode: auto-fill
% fill-column: 80
% End:
